### PR TITLE
Comments: Add HTML toolbar to the edit textarea (2nd attempt)

### LINKED ISSUE #21663: Comments: add more formatting options in edit mode
We currently provide folks a plain text area, let's also add some formatting helpers like we do in wp-admin:

screen shot 2018-01-18 at 2 13 03 pm

screen shot 2018-01-18 at 2 18 39 pm

### PR DESCRIPTION
**Pull request by Copons**
Fix #21663
Supersede #22040 

Add an HTML toolbar to help adding quick tags when editing comments in the Comments Management section.

screen shot 2018-02-01 at 12 07 16

All props for this new take go to @dmsnell!

## Testing instructions

Open `/comments` and click on Edit on some comments.
Use the new HTML toolbar to open and close tags.
Make sure the inserted tags feel "right" (e.g. the expected newlines or indents).

### Testing Notes

I've tried my best to fix all the issues reported in the previous flow patrol call for testing (p6jOYE-1ww-p2), but it's still possible some of them slipped through the cracks.
Either way, this is a list of what this PR **won't** attempt to fix.

- The toolbar is only aware of the toolbar itself.
This means that an open tag manually written, won't be closed when clicking on its button.
This also mean that the close tags button won't close any tags manually written (or added via the toolbar during a previous edit).
This is what happens in wp-admin, but also not the best behaviour possible. Though, changes might be too complicated, likely overkill for this use case. Let's leave thinking about improving for another time. 🙂 

- "If I don't close an HTML tag, it isn't auto-closed for me. (on save)" (reported by @rachelmcr).
For now let's try to have the same behaviour of wp-admin, and then iterate on top of that.
This kind of auto-closing could make use of the internal open tags tracker or of a regular expressions. Either can be done in a separate PR.

- "When adding not closed tag it get's URL encoded in Edit field: `<a` -> `&lt;a`" (reported by @brbrr)
This should indeed be the expected outcome.
The backend parse the comment and attempt to sanitize it. `<a` by itself is not necessarily a malformed tag, but it can be a "less then a" expression.

- The toolbar is only on the Edit textarea. It will be added to the Reply textarea as well in a follow up PR. We'll think about a Preview too in the future. (cc @hoverduck @brbrr)

## Dev Notes

This approach does two things differently than the post editor HTML toolbar.

- It builds the tags by taking advantage of the `Document.createElement`, `Element.setAttribute`, and `Element.outerHTML` methods, instead of manually creating the strings.
This way, we get to both decrease the file size and complexity, and also minimize the human error chances.
(Again, all props to @dmsnell for the clever solution)

- Instead of jumping through multiple hoops to make `document.execCommand( 'insertText' )` (needed for the undo history) behave in all browsers as it does in Chrome, I've just decided to throw in the towel, and embrace the incompatibilities.

### Chrome
It uses `execCommand`, which triggers the textarea `onChange`, which updates the comment content in the component internal state.
Inserted tags are part of the undo history.
All is well in the world.

### Firefox, IE11, Edge
`execCommand` doesn't work, doesn't trigger the textarea `onChange`, and doesn't update the comment content in the component internal state.
As far as I know, there is just no simple way to keep the undo history.
So, we just manually trigger the textarea `onChange`, which updates the content in state, which in turn updates it in the textarea itself.
Then we pass a callback to the `setState`, which sets the cursor position appropriately (usually right after the inserted content), and finally gives the focus back to the textarea.

### CODE REVIEW

File: client/my-sites/comments/comment/comment-html-editor.jsx [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+		const closer =
+			fragments[ 1 ] + ( options.newLineAfter ? '\n' : '' ) + ( options.paragraph ? '\n\n' : '' );
+		const inner = options.text || this.splitSelectedContent().inner;
+
+		if ( inner.length || options.alsoClose ) {
+			return this.insertContent( opener + inner + closer, options.adjustCursorPosition );
+		}
+
+		if ( ! options.selfClosed && this.isTagOpen( tag ) ) {
+			this.setState( ( { openTags } ) => ( { openTags: filter( openTags, tag ) } ) );
```
Comment (dmsnell) [inline client/my-sites/comments/comment/comment-html-editor.jsx]: careful! we're working with a tree data structure (HTML/DOM) not a list

```js
<ul><li>outer<ul><li>inner</li> // oops wiped out the outer `li` too
```

Comment (Copons) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: I don't think I got what you mean.

`state.openTags` is a simple array that gets populated whenever you open a tag **via the toolbar**.
e.g. you click on the [B] button, and `openTags.concat( 'b' )`

`openTags` then is used to naively check if a button should open or close a tag.
e.g. `<button>{ this.isTagOpen( 'b' ) ? '/b' : 'b' }</button>`

when you close the tag (again: only **via the toolbar**), we simply remove it from the list.
e.g `filter( openTags, tag => 'b' === tag )`

Then, if there are open tags, the [CLOSE TAGS] button is enabled, and on click it simply insert the close tag for every element in the list.

If we open a tag via the toolbar and then we close it manually by writing `</tag>`, `state.openTags` goes "out of sync" with the content.

So, in other words, `state.openTags` doesn't do anything to the text.

---

This is exactly the same behaviour of the post editor HTML toolbar, and AFAIK also of wp-admin HTML toolbars.

If you have a better approach (and I'm sure there are, but all quite complicated from a RegExp standpoint), I'd suggest to ship this PR and improve after.

If you're pointing out a bug, could you please explain it more? 😄

Comment (dmsnell) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: ah. in other words, it's not our open/close code that's broken, it's the toolbar which doesn't express the capabilities of HTML…

in that case also check out `lodash.union` and `lodash.difference` for a nice way to add some semantics to the operations of adding and removing an item from a set. there's a native `Set()` in JavaScript built for this but it's not immutable sadly

Comment (Copons) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: Well, we're not creating a WYSIWYG editor, but really just a text editor with some helper buttons.
I've tried with some regex magic to keep track of the open/closed tags, but my skills (Google-foo included) are not nearly enough.

I guess I could try to create some sort of tree structure that mirrors the comment content DOM, but I'd say that's hella overkill for this purpose.

I like the idea of using `Set` though!

Comment (dmsnell) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: well if we immediately switch the button to the tag closer then we're bound by the UI. I think it's okay though. like you said, they are just helpers.

RegExps will fail us - they are guaranteed to.

Comment (gwwar) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: So in a follow up I'd recommend using [DomParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) to do this for us if it's available:

screen shot 2018-02-15 at 10 58 53 am

File: client/my-sites/comments/comment/comment-html-editor.jsx [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+import PropTypes from 'prop-types';
+import { localize } from 'i18n-calypso';
+import classNames from 'classnames';
+import { delay, each, filter, get, map, reduce } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import AddImageDialog from 'post-editor/editor-html-toolbar/add-image-dialog';
+import AddLinkDialog from 'post-editor/editor-html-toolbar/add-link-dialog';
```
Comment (gwwar) [inline client/my-sites/comments/comment/comment-html-editor.jsx]: We're still +50% on comment bundle size. @Copons, as I mentioned before, let's ignore the dialogs for now, and simply insert a tag.

screen shot 2018-02-15 at 11 01 18 am

Comment (gwwar) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: The dialogs and even this toolbar component I think are good candidates to async load in. We can of course try to do that as part of follow up PRs

File: client/my-sites/comments/comment/style.scss [hunk 3]
```diff
@@ -492,6 +493,89 @@
...
+	text-decoration: underline;
+}
+.comment__html-toolbar-button-code {
+	font-family: $code;
+}
+.comment__html-toolbar-button-close-tags.button[disabled]:hover {
+	color: lighten($gray, 30%);
+}
+
+.editor-html-toolbar__dialog {
```
Comment (gwwar) [inline client/my-sites/comments/comment/style.scss]: We'll likely delete this later, but is this meant as an override for this section only? The current rule would modify styles for every `editor-html-toolbar__dialog` class

File: client/my-sites/comments/comment/comment-html-editor.jsx (hunk 1 above)
Comment (dmsnell) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: Thanks @gwwar - that was more helpful than my response. `DOMParser` should be available in every environment we support.

File: client/my-sites/comments/comment/comment-html-editor.jsx (hunk 2 above)
Comment (Copons) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: Sure, you're right, I totally forgot about that removing the dialogs. 🤦‍♂️

File: client/my-sites/comments/comment/style.scss (hunk 3 above)
Comment (Copons) [reply client/my-sites/comments/comment/style.scss]: This is actually a funnier oversight.
The post editor toolbar's CSS is loaded only with the post editor bundle, so by opening Calypso straight into comments, the dialogs were un-styled.
So I just copied the CSS over to comments, duplicating it without thinking twice.

Good point though: when I'll remove the dialogs I'll also clean up the CSS.
Thanks for pointing it out, I'd have likely missed it. 🙇

**Comment by Copons** [comment]
@gwwar Last changes are pending cross-browser tests, but these are the new bundle sizes:

screen shot 2018-02-16 at 13 15 35

**Comment by kwight** [comment]
Gave things another spin on Mac 'n' Chrome; tried adding and closing tags on mobile and desktop, plus used the new dialogs (adding a link and an image). Everything worked as expected 👍 And that new bundle difference is 🔥

**Comment by gwwar** [comment]
I'm okay with a 5% jump, as long as we follow up with some async bundle 🔍 improvements later. 😄

File: client/my-sites/comments/comment/comment-html-editor/add-image-dialog.jsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+	setImageUrl = event => this.setState( { imageUrl: event.target.value } );
+
+	closeDialog = () => {
+		this.setState( {
+			imageAlt: '',
+			imageTitle: '',
+			imageUrl: '',
+		} );
+		this.props.onClose();
```
Comment (dmsnell) [inline client/my-sites/comments/comment/comment-html-editor/add-image-dialog.jsx]: it's probably good to pass `this.props.onClose` as the second argument to `this.setState()` because otherwise we're exposing a potential race condition between calling that function and updating state. if we send it as an argument it will only be called after `setState()` has actually updated local state.

```js
this.setState( { imageAlt: '', imageTitle: '', imageUrl: '' }, this.props.onClose );
```

Note that we're not _calling_ `onClose` but rather _passing_ it as a callback.

File: client/my-sites/comments/comment/comment-html-editor/add-link-dialog.jsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+ */
+import Dialog from 'components/dialog';
+import FormCheckbox from 'components/forms/form-checkbox';
+import FormFieldset from 'components/forms/form-fieldset';
+import FormLabel from 'components/forms/form-label';
+import FormTextInput from 'components/forms/form-text-input';
+
+const REGEXP_EMAIL = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i;
+const REGEXP_URL = /^(https?|ftp):\/\/[A-Z0-9.-]+\.[A-Z]{2,4}[^ "]*$/i;
+const REGEXP_STANDALONE_URL = /^(?:[a-z]+:|#|\?|\.|\/)/;
```
Comment (dmsnell) [inline client/my-sites/comments/comment/comment-html-editor/add-link-dialog.jsx]: so it's okay to use these kinds of RegExps but we need to be explicit with their boundaries - at a MINIMUM an explanation of why we're using degraded parsers and where they will fail.

typically I like to make it obvious in the naming…

```js
const BAD_REGEXP_EMAIL = /…/i;
const BAD_REGEXP_URL = /…/i;
const BAD_STANDALONE_URL = /…/;
```

There are URL parsers we can use, including the `url` package which is in use already around the app, but those run into the problem that almost everything is a valid URL in some sense.

It may be helpful to be pretty lenient here. Did these come from somewhere else?

Also, how is a leading dot a URL? (that's the `\.` alternation value in the standalone pattern)

File: client/my-sites/comments/comment/comment-html-editor/add-link-dialog.jsx [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+	setLinkNewTab = event => this.setState( { linkNewTab: event.target.checked } );
+
+	closeDialog = () => {
+		this.setState( {
+			linkNewTab: false,
+			linkText: '',
+			linkUrl: '',
+		} );
+		this.props.onClose();
```
Comment (dmsnell) [inline client/my-sites/comments/comment/comment-html-editor/add-link-dialog.jsx]: see previous note on concurrency

File: client/my-sites/comments/comment/comment-html-editor/index.jsx [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+		this.insertHtmlTag( 'a', attributes, { adjustCursorPosition: -4, alsoClose: true } );
+	};
+
+	insertBlockquoteTag = () => this.insertHtmlTag( 'blockquote', {}, { paragraph: true } );
+
+	insertDelTag = () => this.insertHtmlTag( 'del', { datetime: this.props.moment().format() } );
+
+	insertInsTag = () => this.insertHtmlTag( 'ins', { datetime: this.props.moment().format() } );
+
+	insertImgTag = attributes => this.insertHtmlTag( 'img', attributes, { selfClosed: true } );
```
Comment (dmsnell) [inline client/my-sites/comments/comment/comment-html-editor/index.jsx]: note that as an empty element this needs no `selfClosed` attribute - we should know that inside of `insertHtmlTag` because `fragments[1]` will be undefined.

File: client/my-sites/comments/comment/comment-html-editor/add-link-dialog.jsx (hunk 5 above)
Comment (Copons) [reply client/my-sites/comments/comment/comment-html-editor/add-link-dialog.jsx]: > Did these come from somewhere else?

Yup: https://github.com/Automattic/wp-calypso/blob/master/client/components/tinymce/plugins/wplink/dialog.jsx#L34-L36

These regexes also appears in the post editor HTML toolbar's link dialog.
The idea was to have a behaviour 100% identical to the visual toolbar / TinyMCE link dialog.

I'm not opposed to update/fix them, but I believe we should do it in a follow up PR, tackling all 3 occurrences (MCE plugin, post editor HTML toolbar, comment editor HTML toolbar) at once.

**Comment by vindl** [comment]
I've run into an issue when closing nested tags: 

comment-editor1
I expected that option to close the `b` tag would remain available here. For comparison, this works as expected in `wp-admin` equivalent.
